src/main/java/epam/gym/service/hibernateImpl/TraineeServiceImpl.java

```diff
@@ -0,0 +1,203 @@
+            logger.info("Deleted trainee with ID: {}", id);
+        } else {
+            logger.error("Authentication failed for user: {}", username);
+            throw new AuthenticationException("Authentication failed for user: " + username);
+        }
+    }
+
+    @Override
+    public Optional<Trainee> selectTraineeByUsername(String username) {
+        Optional<Trainee> traineeOpt = traineeDao.selectTraineeByUsername(username);
+        if (traineeOpt.isPresent()) {
+            logger.info("Found trainee with username: {}", username);
+        } else {
+            logger.warn("Failed to find trainee with username: {}", username);
+        }
+        return traineeOpt;
+    }
+
+    @Override
+    public Optional<Trainee> changePassword(Long id, String username, String oldPassword, String newPassword) throws AuthenticationException {
+        if (authenticationService.authenticate(id, username, oldPassword)) {
+            Optional<Trainee> updatedTrainee = traineeDao.changePassword(id, newPassword);
+            if (updatedTrainee.isPresent()) {
+                logger.info("Password changed successfully for trainee with ID: {}", id);
+            } else {
+                logger.warn("Failed to change password. Trainee with ID: {} not found.", id);
+            }
+            return updatedTrainee;
+        } else {
+            logger.error("Authentication failed for user: {}", username);
+            throw new AuthenticationException("Authentication failed for user: " + username);
+        }
+    }
+
+
+    @Override
+    public void activateTrainee(Long id, String username, String password) throws AuthenticationException {
+        if (authenticationService.authenticate(id, username, password)) {
+            Optional<Trainee> traineeOpt = findById(id, username, password);
+            if (traineeOpt.isPresent()) {
+                traineeDao.setActiveStatus(id, true);
+                logger.info("Activated trainee with ID: {}", id);
+            } else {
+                logger.error("Trainee not found with ID: {}", id);
+                throw new IllegalArgumentException("Trainee not found");
+            }
+        } else {
+            logger.error("Authentication failed for user: {}", username);
+            throw new AuthenticationException("Authentication failed for user: " + username);
+        }
+    }
+
+    @Override
+    public void deactivateTrainee(Long id, String username, String password) throws AuthenticationException {
+        if (authenticationService.authenticate(id, username, password)) {
+            Optional<Trainee> traineeOpt = findById(id, username, password);
+            if (traineeOpt.isPresent()) {
+                traineeDao.setActiveStatus(id, false);
+                logger.info("Deactivated trainee with ID: {}", id);
+            } else {
+                logger.error("Trainee not found with ID: {}", id);
+                throw new IllegalArgumentException("Trainee not found");
```

Comment:
Create own exception, EntityNotFoundException for example. IllegalArgumentException is used when bad arguments in method